File: MTools/DataDec.cpp
#include "DataDec.h"
#include "ByteUtil.h"

#include <unistd.h>
#include <cstring>

DataDec::DataDec() {
}


DataDec::DataDec(mbyte *bytes, int bytelen) {
    setData(bytes, bytelen);
}


void DataDec::setData(mbyte *bytes, int bytelen) {
    reset();
    m_byteLen = bytelen;
    m_bytes = bytes;
}

int DataDec::getInt() {
    int val = getInt(index);
    index += 4;
    return val;
}

mlong DataDec::getLong() {
    mlong val = getLong(index);
    index += 8;
    return val;
}

mbyte DataDec::getByte() {
    mbyte val = getByte(index);
    index += 1;
    return val;
}

bool DataDec::getBool() {
    return getByte() > 0;
}

char *DataDec::getStr() {
    int len = getInt();
    if (len > 0 && (index + len) <= m_byteLen) {
        char *str = new char[len + 1];
        memset(str, 0, len);
        memcpy(str, m_bytes + index, len);
        str[len] = '\0';
        index += len;
        return str;
    }
    return nullptr;
}

float DataDec::getFloat() {
    float val = getFloat(index);
    index += 4;
    return val;
}

double DataDec::getDouble() {
    double val = getDouble(index);
    index += 8;
    return val;
}


int DataDec::getCmd() {
    return getInt(0);
}

int DataDec::getCount() {
    return getInt(4);
}

mlong DataDec::getLength() {
    return getLong(8);
}

void DataDec::skip(int off) {
    index = HEADER_LEN + off;
}

void DataDec::reset() {
    index = HEADER_LEN;
}

int DataDec::headerSize() {
    return HEADER_LEN;
}

int DataDec::getInt(int i) {
    if ((i + 4) <= m_byteLen) {
        int val = ByteUtil::bytesToInt(m_bytes, i);
        return val;
    }
    return 0;
}

mlong DataDec::getLong(int i) {
    if ((i + 8) <= m_byteLen) {
        mlong lg = ByteUtil::bytesToLong(m_bytes, i);
        return lg;
    }
    return 0;
}

mbyte DataDec::getByte(int i) {
    if ((i + 1) <= m_byteLen) {
        mbyte b = m_bytes[i];
        return b;
    }
    return 0;
}

char *DataDec::getStr(int i) {
    int len = getInt(i);
    i += 4;
    if (len > 0 && (i + len) <= m_byteLen) {
        char *str = new char[len];
        memcpy(str, m_bytes + i, len);
        str[len] = '\0';
        return str;
    }
    return nullptr;
}

float DataDec::getFloat(int i) {
    return getInt(i) / 1000;
}

double DataDec::getDouble(int i) {
    return getLong(i) / 1000000;
}

File: java/utils/DataDec.java
package connect.utils;

import java.io.UnsupportedEncodingException;


//数据解析类
public class DataDec {
    public static final int HEADER_LEN = 16;
    private byte bytes[] = null;
    private int index = HEADER_LEN;


    public DataDec() {

    }

    public DataDec(byte bs[]) {
        setData(bs);
    }

    public void setData(byte bs[]) {
        reset(); // 重置读取下标
        this.bytes = bs;
    }


    public int getInt() {
        if (bytes != null && bytes.length >= index + 4) {
            int data = ByteUtil.bytesToInt(bytes, (int) index);
            index += 4;
            return data;
        }
        return 0;
    }

    public boolean getBool() {
        if (bytes != null && bytes.length >= index + 1) {
            return bytes[index++] == 1;
        }
        return false;
    }

    public long getLong() {
        if (bytes != null && bytes.length >= index + 8) {
            long data = ByteUtil.bytesToLong(bytes, (int) index);
            index += 8;
            return data;
        }
        return 0;
    }


    public float getFloat() {
        return getInt() / 1000.0f;
    }

    public double getDouble() {
        if (bytes != null && bytes.length >= index + 8) {
            long data = ByteUtil.bytesToLong(bytes, (int) index);
            index += 8;
            return data / 1000000.0d;
        }
        return 0;
    }

    public long getLastLong() {
        long length = getLength();
        if (length > 0) {
            return ByteUtil.bytesToLong(bytes, (int) +HEADER_LEN - 8);
        }
        return 0;
    }


    /**
     * 设置读取偏移
     *
     * @param skip
     */
    public void skip(int skip) {
        index = HEADER_LEN + skip;
    }

    /**
     * 重置读取下标
     */
    public void reset() {
        index = HEADER_LEN;
    }


    /**
     * 获取String
     *
     * @param en
     * @return
     */
    public String getString(String en) {
        int strLen = ByteUtil.bytesToInt(bytes, (int) index);
        index += 4;
        if (bytes != null && bytes.length >= index + strLen) {
            String str = null;
            try {
                str = new String(bytes, (int) index, strLen, en);
            } catch (UnsupportedEncodingException e) {
                e.printStackTrace();
            }
            index += strLen;
            return str;
        }
        return null;
    }

    public String getString() {
        return getString("UTF-8");
    }

    public int getCmd() {
        if (bytes != null && bytes.length >= HEADER_LEN) {
            return ByteUtil.bytesToInt(bytes, 0);
        }
        return -1;
    }

    public int getCount() {
        if (bytes != null && bytes.length >= HEADER_LEN) {
            return ByteUtil.bytesToInt(bytes, 4);
        }
        return -1;
    }

    public long getLength() {
        if (bytes != null && bytes.length >= HEADER_LEN) {
            return ByteUtil.bytesToLong(bytes, 8);
        }
        return -1;
    }

    public int getHeaderSize() {
        return HEADER_LEN;
    }


}
